### PR TITLE
RavenDB-13679 UI for Server-wide Backup

### PR DESCRIPTION
**Pull request by Danielle9897**

### CODE REVIEW

File: src/Raven.Client/ServerWide/Operations/Configuration/ServerWideBackupConfiguration.cs [hunk 1]
```diff
@@ -1,9 +1,31 @@
...
 
 namespace Raven.Client.ServerWide.Operations.Configuration
 {
     public class ServerWideBackupConfiguration : PeriodicBackupConfiguration
     {
         internal static string NamePrefix = "Server Wide Backup";
+        
+        // Properties to be used by the studio server-wide backups list view
+        public OngoingTaskState TaskState { get; set; }
+        public List<string> BackupDestinations { get; set; }
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/ServerWideBackupConfiguration.cs]: Not sure why those are needed here. Even if they are, then I suggest creating a class `ServerWideBackupConfigurationForStudio` (in server project - probably in handler) that inherits from `ServerWideBackupConfiguration` and add it there

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+using Raven.Server.Rachis;
+using Raven.Client.ServerWide.Operations.Configuration;
+using Raven.Server.Documents.PeriodicBackup;
+using Raven.Server.Web.Studio;
+
+namespace Raven.Server.Web.System
+{
+    public class BackupServerWideHandler : RequestHandler
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/BackupServerWideHandler.cs]: rename to `AdminServerWideBackupHandler`

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                
+                context.Write(writer, backupsResult.ToJson());
+                writer.Flush();
+
+                return Task.CompletedTask;
+            }
+        }
+        
+        // Toggle Disable/Enable for specified task
+        [RavenAction("/admin/configuration/serve-wide/backup/state", "POST", AuthorizationStatus.ClusterAdmin)]
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/BackupServerWideHandler.cs]: typo: serve -> server

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        public async Task ToggleTaskState()
+        {
+            var disable = GetBoolValueQueryString("disable") ?? true;
+            var taskName = GetStringQueryString("taskName", required: false);
+           
+            using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
+            {
+                // Get existing task
+                var backups = ServerStore.Cluster.GetServerWideBackupConfigurations(context, taskName);
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/BackupServerWideHandler.cs]: Can you have more than one?

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+            var taskName = GetStringQueryString("taskName", required: false);
+           
+            using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
+            {
+                // Get existing task
+                var backups = ServerStore.Cluster.GetServerWideBackupConfigurations(context, taskName); 
+                ServerWideBackupConfigurationResults backupsResult = new ServerWideBackupConfigurationResults();
+                
+                if (backups.Count() == 0)
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/BackupServerWideHandler.cs]: Use `FirstOrDefault` so you will avoid doing `Count()` then `ToList()[0]`

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs (hunk 4 above)
Comment (ppekrol) [reply src/Raven.Server/Web/System/BackupServerWideHandler.cs]: If yes, then we need to switch all of them, but it should be unique IIRC?

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+                using (var writer = new BlittableJsonTextWriter(context, ResponseBodyStream()))
+                {
+                    var test = new ModifyServerWideBackupResult() 
+                    {
+                        TaskName = taskName,
+                        RaftCommandIndex = newIndex
+                    };
+
+                    HttpContext.Response.StatusCode = (int)HttpStatusCode.OK;
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/BackupServerWideHandler.cs]: OK is the default, you do not need to set this.

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+            await info.UpdateDirectoryResult(databaseName: null, error: pathResult.Error);
+        }
+    }
+    
+    public class ModifyServerWideBackupResult : IDynamicJson 
+    {
+        public string TaskName { get; set; }
+        public long RaftCommandIndex { get; set; }
+
+        public virtual DynamicJsonValue ToJson()
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/BackupServerWideHandler.cs]: does not need to be virtual

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+            }
+        }
+        
+        [RavenAction("/admin/backup-data-directory", "GET", AuthorizationStatus.ClusterAdmin)]
+        public async Task FullBackupDataDirectory()
+        {
+            var path = GetStringQueryString("path", required: true);
+            var requestTimeoutInMs = GetIntValueQueryString("requestTimeoutInMs", required: false) ?? 5 * 1000;
+
+            var pathResult = BackupConfigurationHelper.GetActualFullPath(ServerStore, path);
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/BackupServerWideHandler.cs]: Create a static method that will take path, requestTimeoutInMs and getNodesInfo as a list of arguments and will perform the operation. Use that in both places so we will not have desync in the future between those two endpoints.

File: src/Raven.Studio/typescript/commands/resources/getAllServerWideBackupTasksCommand.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import commandBase = require("commands/commandBase");
+import endpoints = require("endpoints");
+
+class getAllServerWideBackupTasksCommand extends commandBase {
+    constructor() {
```
Comment (ml054) [inline src/Raven.Studio/typescript/commands/resources/getAllServerWideBackupTasksCommand.ts]: no need for constuctor

File: src/Raven.Studio/typescript/commands/resources/getServerWideBackupConfigCommand.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
+import commandBase = require("commands/commandBase");
+import endpoints = require("endpoints");
+
+class getServerWideBackupConfigCommand extends commandBase {
+
+    constructor() {
```
Comment (ml054) [inline src/Raven.Studio/typescript/commands/resources/getServerWideBackupConfigCommand.ts]: no need

File: src/Raven.Studio/typescript/models/database/tasks/ongoingTaskBackupListModel.ts [hunk 11]
```diff
@@ -16,8 +16,10 @@ import shell = require("viewmodels/shell");
 class ongoingTaskBackupListModel extends ongoingTaskListModel {
     private static neverBackedUpText = "Never backed up";
     
-    private static namePrefix = "Server Wide Backup";
-
+    private static serverWideNamePrefixFromServer = "Server Wide Backup";
```
Comment (ml054) [inline src/Raven.Studio/typescript/models/database/tasks/ongoingTaskBackupListModel.ts:19]: why do we have this discrepancy?

File: src/Raven.Studio/typescript/models/database/tasks/ongoingTaskBackupListModel.ts [hunk 12]
```diff
@@ -62,8 +62,17 @@ class ongoingTaskBackupListModel extends ongoingTaskListModel {
...
-        const urls = appUrl.forCurrentDatabase();
-        this.editUrl = urls.editPeriodicBackupTask(this.taskId);
+        this.isServerWide = ko.pureComputed(() => {
+            return this.taskName().startsWith(ongoingTaskBackupListModel.serverWideNamePrefixInStudio);
+        });
+        
+        this.editUrl = ko.pureComputed(()=> {
+             const urls = appUrl.forCurrentDatabase();
+             
+             return this.isServerWide() ?  appUrl.forEditServerWideBackup(this.taskName().replace(ongoingTaskBackupListModel.serverWideNamePrefixInStudio +', ', '')) :
```
Comment (ml054) [inline src/Raven.Studio/typescript/models/database/tasks/ongoingTaskBackupListModel.ts]: can we be consistent with server?

File: src/Raven.Studio/typescript/viewmodels/manage/editServerWideBackup.ts [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+                    .done((backupsList: Raven.Server.Web.System.ServerWideBackupConfigurationResults) => {
+                        if (backupsList.Results.length) {
+                            const backupTask = backupsList.Results[0];
+                            if (this.serverConfiguration().LocalRootPath && backupTask.LocalSettings.FolderPath && backupTask.LocalSettings.FolderPath.startsWith(this.serverConfiguration().LocalRootPath)) {
+                                backupTask.LocalSettings.FolderPath = backupTask.LocalSettings.FolderPath.substr(this.serverConfiguration().LocalRootPath.length);
+                            }
+
+                            this.configuration(new periodicBackupConfiguration(backupTask, this.serverConfiguration(), false, true)); 
+                            deferred.resolve();
+                        }
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/manage/editServerWideBackup.ts]: Please format else correctly: 
} else {

File: src/Raven.Studio/typescript/viewmodels/manage/editServerWideBackup.ts [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+                    if (!encryptionSettings.key()) {
+                        return this.encryptionSection.generateEncryptionKey()
+                            .done(() => {
+                                encryptionSettings.dirtyFlag().reset();
+                            });
+                    }
+                });
+        };
+
+        return $.when<any>(this.loadServerSideConfiguration())
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/manage/editServerWideBackup.ts]: no need to $.when here - you have single task to await.

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs (hunk 4 above)
Comment (Danielle9897) [reply src/Raven.Server/Web/System/BackupServerWideHandler.cs]: 1. Yes - there is just '**one**' - I will refactor this.
2. But, actually, we need to add server code that prevents creating a **new** task with an existing task name - as it is currently Not handled and new task with an existing name overrides the existing task

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs (hunk 6 above)
Comment (Danielle9897) [reply src/Raven.Server/Web/System/BackupServerWideHandler.cs]: OK removed

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs (hunk 5 above)
Comment (Danielle9897) [reply src/Raven.Server/Web/System/BackupServerWideHandler.cs]: done

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs (hunk 3 above)
Comment (Danielle9897) [reply src/Raven.Server/Web/System/BackupServerWideHandler.cs]: done

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs (hunk 7 above)
Comment (Danielle9897) [reply src/Raven.Server/Web/System/BackupServerWideHandler.cs]: done

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs (hunk 2 above)
Comment (Danielle9897) [reply src/Raven.Server/Web/System/BackupServerWideHandler.cs]: done

File: src/Raven.Client/ServerWide/Operations/Configuration/ServerWideBackupConfiguration.cs (hunk 1 above)
Comment (Danielle9897) [reply src/Raven.Client/ServerWide/Operations/Configuration/ServerWideBackupConfiguration.cs]: done

File: src/Raven.Server/Web/System/BackupServerWideHandler.cs (hunk 8 above)
Comment (Danielle9897) [reply src/Raven.Server/Web/System/BackupServerWideHandler.cs]: done

File: src/Raven.Studio/typescript/commands/resources/getAllServerWideBackupTasksCommand.ts (hunk 9 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/commands/resources/getAllServerWideBackupTasksCommand.ts]: done

File: src/Raven.Studio/typescript/commands/resources/getServerWideBackupConfigCommand.ts (hunk 10 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/commands/resources/getServerWideBackupConfigCommand.ts]: done

File: src/Raven.Studio/typescript/models/database/tasks/ongoingTaskBackupListModel.ts (hunk 11 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/models/database/tasks/ongoingTaskBackupListModel.ts:19]: I cannot change this prefix on the server side - too late for that - 
so I can only manipulate the way it looks in the UI

File: src/Raven.Studio/typescript/viewmodels/manage/editServerWideBackup.ts (hunk 13 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/manage/editServerWideBackup.ts]: done

File: src/Raven.Studio/typescript/viewmodels/manage/editServerWideBackup.ts (hunk 14 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/manage/editServerWideBackup.ts]: done

File: src/Raven.Studio/typescript/models/database/tasks/ongoingTaskBackupListModel.ts (hunk 12 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/models/database/tasks/ongoingTaskBackupListModel.ts]: As discussed:
1. Ongoing-tasks-list-view will show: `SERVER WIDE BACKUP, <TASK NAME> `
2. Server-Wide-backups-list-view will show just:  `<TASK NAME>`

**Pull request merged**